### ISSUE TITLE
onprogress issue while converting to HLS

### ISSUE THREAD

**Issue by Babiute** [protonemedia/laravel-ffmpeg#294]
Hey Pascal.

Look that. While Im converting a video to HLS, the onprogress callback is not working properly.

If I remove it, everything works fine, but if not, it throws error.

RN I'm without any ideas about how to solve it or why its happening.

The code is the bellow

```
        $stream = new Video();

        FFMpeg::fromDisk('from')
            ->open('file_name')
            ->exportForHLS()
            ->toDisk('to')
            ->addFormat((new X264('aac'))
                ->setKiloBitrate(500)
                ->setAudioKiloBitrate(126)
            ->onProgress(function ($percentage, $remaining, $rate) use ($stream) { //this is the line 77
                $stream->progress= $percentage;
                $stream->save();
            })
            ->save('my_file.m3u8');
```

The error is:

```
[2021-03-02 12:15:31] local.ERROR: Too few arguments to function App\Extras\Video\VideoStream::App\Extras\Video\{closure}(), 1 passed and exactly 3 expected {"exception":"[object] (ArgumentCountError(code: 0): Too few arguments to function App\\Extras\\Video\\VideoStream::App\\Extras\\Video\\{closure}(), 1 passed and exactly 3 expected at C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\app\\Extras\\Video\\VideoStream.php:77)
[stacktrace]
#0 [internal function]: App\\Extras\\Video\\VideoStream->App\\Extras\\Video\\{closure}(100)
#1 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\pbmedia\\laravel-ffmpeg\\src\\Exporters\\MediaExporter.php(207): call_user_func(Object(Closure), 100)
#2 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\pbmedia\\laravel-ffmpeg\\src\\Exporters\\MediaExporter.php(149): ProtoneMedia\\LaravelFFMpeg\\Exporters\\MediaExporter->saveWithMappings()
#3 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\pbmedia\\laravel-ffmpeg\\src\\Exporters\\HLSExporter.php(304): ProtoneMedia\\LaravelFFMpeg\\Exporters\\MediaExporter->save()
#4 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Collections\\Traits\\EnumeratesValues.php(692): ProtoneMedia\\LaravelFFMpeg\\Exporters\\HLSExporter->ProtoneMedia\\LaravelFFMpeg\\Exporters\\{closure}(Object(Illuminate\\Support\\Collection))
#5 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\pbmedia\\laravel-ffmpeg\\src\\Exporters\\HLSExporter.php(318): Illuminate\\Support\\Collection->pipe(Object(Closure))
#6 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\app\\Extras\\Video\\VideoStream.php(81): ProtoneMedia\\LaravelFFMpeg\\Exporters\\HLSExporter->save('HCbpcyT3zpPEUoq...')
#7 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\app\\Extras\\Video\\VideoStream.php(55): App\\Extras\\Video\\VideoStream->gerarSteram()
#8 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\app\\Jobs\\CriarVideoStream.php(58): App\\Extras\\Video\\VideoStream->stream(Object(App\\Models\\Video))
#9 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Container\\BoundMethod.php(36): App\\Jobs\\CriarVideoStream->handle()
#10 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Container\\Util.php(40): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
#11 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Container\\BoundMethod.php(93): Illuminate\\Container\\Util::unwrapIfClosure(Object(Closure))
#12 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Container\\BoundMethod.php(37): Illuminate\\Container\\BoundMethod::callBoundMethod(Object(Illuminate\\Foundation\\Application), Array, Object(Closure))
#13 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Container\\Container.php(610): Illuminate\\Container\\BoundMethod::call(Object(Illuminate\\Foundation\\Application), Array, Array, NULL)
#14 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Bus\\Dispatcher.php(128): Illuminate\\Container\\Container->call(Array)
#15 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Pipeline\\Pipeline.php(128): Illuminate\\Bus\\Dispatcher->Illuminate\\Bus\\{closure}(Object(App\\Jobs\\CriarVideoStream))
#16 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Pipeline\\Pipeline.php(103): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(App\\Jobs\\CriarVideoStream))
#17 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Bus\\Dispatcher.php(132): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))
#18 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\CallQueuedHandler.php(118): Illuminate\\Bus\\Dispatcher->dispatchNow(Object(App\\Jobs\\CriarVideoStream), false)
#19 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Pipeline\\Pipeline.php(128): Illuminate\\Queue\\CallQueuedHandler->Illuminate\\Queue\\{closure}(Object(App\\Jobs\\CriarVideoStream))
#20 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Pipeline\\Pipeline.php(103): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(App\\Jobs\\CriarVideoStream))
#21 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\CallQueuedHandler.php(120): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))
#22 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\CallQueuedHandler.php(70): Illuminate\\Queue\\CallQueuedHandler->dispatchThroughMiddleware(Object(Illuminate\\Queue\\Jobs\\DatabaseJob), Object(App\\Jobs\\CriarVideoStream))
#23 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\Jobs\\Job.php(98): Illuminate\\Queue\\CallQueuedHandler->call(Object(Illuminate\\Queue\\Jobs\\DatabaseJob), Array)
#24 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\Worker.php(406): Illuminate\\Queue\\Jobs\\Job->fire()
#25 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\Worker.php(356): Illuminate\\Queue\\Worker->process('database', Object(Illuminate\\Queue\\Jobs\\DatabaseJob), Object(Illuminate\\Queue\\WorkerOptions))
#26 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\Worker.php(158): Illuminate\\Queue\\Worker->runJob(Object(Illuminate\\Queue\\Jobs\\DatabaseJob), 'database', Object(Illuminate\\Queue\\WorkerOptions))
#27 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\Console\\WorkCommand.php(116): Illuminate\\Queue\\Worker->daemon('database', 'criar_stream', Object(Illuminate\\Queue\\WorkerOptions))
#28 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Queue\\Console\\WorkCommand.php(100): Illuminate\\Queue\\Console\\WorkCommand->runWorker('database', 'criar_stream')
#29 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Container\\BoundMethod.php(36): Illuminate\\Queue\\Console\\WorkCommand->handle()
#30 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Container\\Util.php(40): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
#31 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Container\\BoundMethod.php(93): Illuminate\\Container\\Util::unwrapIfClosure(Object(Closure))
#32 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Container\\BoundMethod.php(37): Illuminate\\Container\\BoundMethod::callBoundMethod(Object(Illuminate\\Foundation\\Application), Array, Object(Closure))
#33 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Container\\Container.php(610): Illuminate\\Container\\BoundMethod::call(Object(Illuminate\\Foundation\\Application), Array, Array, NULL)
#34 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Console\\Command.php(136): Illuminate\\Container\\Container->call(Array)
#35 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\symfony\\console\\Command\\Command.php(256): Illuminate\\Console\\Command->execute(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Illuminate\\Console\\OutputStyle))
#36 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Console\\Command.php(121): Symfony\\Component\\Console\\Command\\Command->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Illuminate\\Console\\OutputStyle))
#37 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\symfony\\console\\Application.php(971): Illuminate\\Console\\Command->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#38 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\symfony\\console\\Application.php(290): Symfony\\Component\\Console\\Application->doRunCommand(Object(Illuminate\\Queue\\Console\\WorkCommand), Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#39 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\symfony\\console\\Application.php(166): Symfony\\Component\\Console\\Application->doRun(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#40 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Console\\Application.php(93): Symfony\\Component\\Console\\Application->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#41 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\vendor\\laravel\\framework\\src\\Illuminate\\Foundation\\Console\\Kernel.php(129): Illuminate\\Console\\Application->run(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#42 C:\\Users\\Administrador\\Desktop\\sesut\\vijeo\\trunk\\vijeo\\artisan(37): Illuminate\\Foundation\\Console\\Kernel->handle(Object(Symfony\\Component\\Console\\Input\\ArgvInput), Object(Symfony\\Component\\Console\\Output\\ConsoleOutput))
#43 {main}
"} 
```

**Comment by pascalbaljet** [comment]
You've made a typo around `->onProgress`. You should call the `onProgress` method on the exporter, not on the format:

```php
FFMpeg::fromDisk('from')
    ->open('file_name')
    ->exportForHLS()
    ->toDisk('to')
    ->addFormat(
        (new X264('aac'))
            ->setKiloBitrate(500)
            ->setAudioKiloBitrate(126)
    )
    ->onProgress(function ($percentage, $remaining, $rate) use ($stream) { 
        $stream->progress = $percentage;
        $stream->save();
    })
    ->save('my_file.m3u8');
````

**Comment by Babiute** [comment]
Hey @pascalbaljet Thx for your replay.

As there are some validations before executing the command to determine the bitrate, I ended up excluding this part and entering the values ​​directly because the validations were not necessary here. With that I ended up misleading you.
I completely copied your code, I made sure there was no typo, but the error persists.
The faithfully executed code was yours and the error continues to be the one shown above.

**Comment by Artem-Schander** [comment]
Hi, i don't know if it's related, but in my case the onProgress callback stopped working. There is no error.
I'm also converting to HLS

**Comment by PaulMorel** [comment]
I've hit this problem myself. What I found is that in the `onProgress` callback, when `$percentage` gets to 100 at the end of the transcoding process, `$remaining` and `$rate` aren't set anymore. The callback, expecting 3 arguments only gets 1. My workaround was to set default values forwhenever `$remaining` and `$rate` are unset.

Here's an example of my usage:
```php
FFMpeg::fromDisk('root')
->open($temporaryDirectory->path($filename))
->fromDisk('private')
->open($music->url)
->export()
->addFormatOutputMapping((new X264())->setKiloBitrate(4000), Media::make('spaces', "public/videos/{$filename}"), ['0:v', '1:a'])
->withVisibility('public')
->onProgress(function ($percentage, $remaining = 0, $rate = 0) {
    Log::info('Export', compact('percentage', 'remaining', 'rate'));
    $this->video->update([
        'processing_progress' => $percentage
    ]);
})->save();
```

**Comment by pascalbaljet** [comment]
Fixed in v7.7.1.
